hack/generated/controllers/generic_controller.go

```diff
@@ -0,0 +1,749 @@
+		if err := register(mgr, applier, obj, log, options); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errs
+}
+
+// TODO: is this comment still correct...?
+// register takes a manager and a struct describing how to instantiate
+// controllers for various types using a generic reconciler function. Only one
+// type (using For) may be directly watched by each controller, but zero, one or
+// many Owned types are acceptable. This setup allows reconcileFn to have access
+// to the concrete type defined as part of a closure, while allowing for
+// independent controllers per GVK (== better parallelism, vs 1 controllers
+// managing many, many List/Watches)
+func register(mgr ctrl.Manager, applier armclient.Applier, obj runtime.Object, log logr.Logger, options controller.Options) error {
+	v, err := conversion.EnforcePtr(obj)
+	if err != nil {
+		return err
+	}
+
+	t := v.Type()
+	controllerName := fmt.Sprintf("%sController", t.Name())
+
+	// Use the provided GVK to construct a new runtime object of the desired concrete type.
+	gvk, err := apiutil.GVKForObject(obj, mgr.GetScheme())
+	if err != nil {
+		return err
+	}
+	log.V(4).Info("Registering", "GVK", gvk)
+
+	// TODO: Do we need to do this?
+	//if err := mgr.GetFieldIndexer().IndexField(obj, "status.id", func(obj runtime.Object) []string {
+	//	unObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	//	if err != nil {
+	//		return []string{}
+	//	}
+	//
+	//	id, ok, err := unstructured.NestedString(unObj, "status", "id")
+	//	if err != nil || !ok {
+	//		return []string{}
+	//	}
+	//
+	//	return []string{id}
+	//}); err != nil {
+	//	return fmt.Errorf("unable to setup field indexer for status.id of %v with: %w", gvk, err)
+	//}
+
+	reconciler := &GenericReconciler{
+		Client:    mgr.GetClient(),
+		ARMClient: applier,
+		Scheme:    mgr.GetScheme(),
+		Name:      t.Name(),
+		Log:       log.WithName(controllerName),
+		Recorder:  mgr.GetEventRecorderFor(controllerName),
+		GVK:       gvk,
+	}
+
+	ctrlBuilder := ctrl.NewControllerManagedBy(mgr).
+		For(obj).
+		WithOptions(options)
+
+	// TODO: I think we need to do the below, but skipping it for now...
+	//if metaObj, ok := obj.(azcorev1.MetaObject); ok {
+	//	trls, err := xform.GetTypeReferenceData(metaObj)
+	//	if err != nil {
+	//		return fmt.Errorf("unable get type reference data for obj %v with: %w", metaObj, err)
+	//	}
+	//
+	//	for _, trl := range trls {
+	//		if trl.IsOwned {
+	//			gvk := schema.GroupVersionKind{
+	//				Group:   trl.Group,
+	//				Version: "v1",
+	//				Kind:    trl.Kind,
+	//			}
+	//			ownedObj, err := mgr.GetScheme().New(gvk)
+	//			if err != nil {
+	//				return fmt.Errorf("unable to create GVK %v with: %w", gvk, err)
+	//			}
+	//
+	//			reconciler.Owns = append(reconciler.Owns, ownedObj)
+	//			ctrlBuilder.Owns(ownedObj)
+	//		}
+	//	}
+	//}
+
+	c, err := ctrlBuilder.Build(reconciler)
+	if err != nil {
+		return fmt.Errorf("unable to build controllers / reconciler with: %w", err)
```

Comment:
yeah I had tried to convert most to errors (I copied from DavidJ originally and nothing was using that), but I must've missed a few. Have gone through and fixed it so that there is no usage of `fmt.Errorf`.